File: c++/Day2.cpp
#include "tl/getlines.hpp"
#include "tl/to.hpp"
#include "bits/stdc++.h"

int main() {
    std::map<std::string, int> mp;
    std::map<std::string, std::string> mp2;

    mp["A X"] = 4; mp["A Y"] = 8; mp["A Z"] = 3;
    mp["B X"] = 1; mp["B Y"] = 5; mp["B Z"] = 9;
    mp["C X"] = 7; mp["C Y"] = 2; mp["C Z"] = 6;

    mp2["A X"] = "A Z"; mp2["A Y"] = "A X"; mp2["A Z"] = "A Y";
    mp2["B X"] = "B X"; mp2["B Y"] = "B Y"; mp2["B Z"] = "B Z";
    mp2["C X"] = "C Y"; mp2["C Y"] = "C Z"; mp2["C Z"] = "C X";

    int ttl1 {0}, ttl2 {0};
    std::ifstream file("Advent_2022_Files/Day2.txt");
    auto lines = tl::views::getlines(file) | tl::to<std::vector>();

    for (auto i : lines) {
        ttl1 += mp[i];
        ttl2 += mp[mp2[i]];
    }

    printf("Part 1: %d\nPart 2: %d", ttl1, ttl2);
}

File: rust/day2.rs
use std::{collections::HashMap, fs::read_to_string};

fn main() {
    let input = read_to_string("../inputs/day2.txt").unwrap();
    solution(&input);
}

fn solution(input: &str) {
    let rps = HashMap::from([
        ("A X", [4, 3]),
        ("B X", [1, 1]),
        ("C X", [7, 2]),
        ("A Y", [8, 4]),
        ("B Y", [5, 5]),
        ("C Y", [2, 6]),
        ("A Z", [3, 8]),
        ("B Z", [9, 9]),
        ("C Z", [6, 7]),
    ]);

    let pt1 = input
        .split("\r\n\r\n")
        .flat_map(|elf| elf.lines())
        .map(|s| rps[s][0])
        .sum::<u32>();

    let pt2 = input
        .split("\r\n\r\n")
        .flat_map(|elf| elf.lines())
        .map(|s| rps[s][1])
        .sum::<u32>();

    println!("Part 1: {pt1}\nPart 2: {pt2}")
}
